src/utils/communication_buffer.hpp

```diff
@@ -0,0 +1,418 @@
+//========================================================================================
+// (C) (or copyright) 2020-2022. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef UTILS_COMMUNICATION_BUFFER_HPP_
+#define UTILS_COMMUNICATION_BUFFER_HPP_
+
+#include <iostream>
+#include <limits>
+#include <memory>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+
+#include "utils/mpi_types.hpp"
+
+#ifdef MPI_PARALLEL
+#include <mpi.h>
+
+#define request_t MPI_Request
+#define comm_t MPI_Comm
+#else
+#define request_t int
+#define comm_t int
+#endif
+
+namespace parthenon {
+
+//             Read    Write
+//    stale:             X
+// sending*:
+// received:     X
+enum class BufferState { stale, sending, sending_null, received, received_null };
+
+enum class BuffCommType { sender, receiver, both, sparse_receiver };
+
+enum class BoundaryType : int { local, nonlocal, any, reflux_send, reflux_recv };
```

Comment:
After having read `reflux` quite often now, I'm curious what this stands for?\
I first thought it's an abbreviation for `receive flux`, but then `reflux_recv` does not make sense anymore.

Reply:
reflux is synonymous with flux correction (it is what they call it in the einstein toolkit, so sometimes I use that nomenclature without thinking about it). I will change it to `flxcorrect` since it is confusing.

Reply:
I replaced reflux with flxcor everywhere for consistency.